contributing/architecture-eval-lifecycle.md

```diff
@@ -0,0 +1,326 @@
+    activate broker
+    broker ->> leaderFSM: complete Evaluation
+    activate leaderFSM
+    leaderFSM ->> workerFSM: replicate Evaluation to followers
+    workerFSM -->> leaderFSM: ok
+    leaderFSM -->> broker: ok
+    deactivate leaderFSM
+    broker -->> sched: ok
+    deactivate broker
+    deactivate sched
+```
+
+## Deployment Watcher
+
+As noted under Scheduling above, a Deployment is created for service jobs. A
+deployment watcher runs on the leader. Its job is to watch the state of
+Allocations being placed for a given job version and to emit new Evaluations so
+that more Allocations for that job can be created.
+
+The "deployments watcher" (plural) makes a blocking query for Deployments and
+spins up a new "deployment watcher" (singular) for each one. That goroutine will
+live until its Deployment is complete or failed.
+
+The deployment watcher makes blocking queries on Allocation health and its own
+Deployment (which can be canceled or paused by a user). When there's a change in
+any of those states, it compares the current state against the [`update`][]
+block and the timers it maintains for `min_healthy_time`, `healthy_deadline`,
+and `progress_deadline`. It then updates the Deployment state and creates a new
+Evaluation if the current step of the update is complete.
+
+The diagram below shows deployments from a high level. Note that Deployments do
+not themselves create Allocations -- they create Evaluations and then the
+schedulers process those as they do normally.
+
+```mermaid
+sequenceDiagram
+
+    participant leaderRpc as Leader RPC
+    participant leaderFSM as Leader State Store
+    participant dw as Deployment Watcher
+    participant workerFSM as Worker State Store
+
+    dw ->> leaderFSM: blocking query for new Deployments
+    activate dw
+    leaderFSM -->> dw: new Deployment
+    dw -->> dw: start watcher
+
+    dw ->> leaderFSM: blocking query for Allocation health
+    leaderFSM -->> dw: Allocation health updates
+    dw ->> dw: next step?
+
+    Note right of dw: Update state and create evaluations for next batch...
+    Note right of dw: Or fail the Deployment and update state
+
+    dw ->> leaderRpc: Deployment.Upsert + Evaluation.Upsert
+    activate leaderRpc
+    leaderRpc ->> leaderFSM: write Deployment and Evaluations
+    activate leaderFSM
+    leaderFSM ->> workerFSM: replicate Deployment + Evaluation to followers
+    workerFSM -->> leaderFSM: ok
+    leaderFSM -->> leaderRpc: ok
+    deactivate leaderFSM
+    leaderRpc -->> dw: ok
+    deactivate leaderRpc
+    deactivate dw
+
+```
+
+## Client Allocs
+
+Once the plan applier has persisted Allocations to the state store (with an
+associated Node ID), they become available to get placed on the client. Clients
+_pull_ new allocations (and changes to allocations), so a new allocation will be
+in the `pending` state until it's been pulled down by a Client and the
+allocation has been instantiated.
```

Comment:
Do clients report allocation status back to the servers in the heartbeat messages?

Reply:
It's a separate message `Node.UpdateAlloc` but we should include that here so that it's obvious how the deployment watcher gets its source of information.

Reply:
Done